torch/fx/passes/runtime_assert.py

```diff
@@ -85,60 +93,140 @@ def insert_deferred_runtime_asserts(
         free_symbols,
         InnerTensorKey,
     )
-    from torch.utils._sympy.interp import sympy_interp
     from torch.utils._sympy.numbers import int_oo
     from torch.utils._sympy.reference import PythonReferenceAnalysis
 
     # TODO: Request simplification on runtime asserts before emitting them
     ras_by_symbol = shape_env.deferred_runtime_asserts.copy()
     graph = gm.graph
-
-    if not any(ras for ras in ras_by_symbol.values()):
-        return
-
     graph_code_log.debug(
         "%s",
         lazy_format_graph_code(
             f"pre insert_deferred_runtime_asserts {name}", gm, colored=True
         ),
     )
 
-    # deduplicate unassociated runtime assertions
-    # we could do better, some guards might be redundant,
-    # e.g. Eq(s0, 4) & Eq(2*s0, 8)
-    # but unclear how to handle all of that right now.
-    # TODO(pianpwk): better way of doing this
-    new_ras = []
-    ras_exprs: Set[sympy.Expr] = set()
-    for ras in ras_by_symbol.pop(None, []):  # type: ignore[call-overload]
-        if ras.expr not in ras_exprs:
-            new_ras.append(ras)
-            ras_exprs.add(ras.expr)
-    ras_by_symbol[None] = new_ras  # type: ignore[index]
-
     # We are going to mutate the dict
     symbol_to_proxy: Dict[sympy.Symbol, fx.Proxy] = {}
     placeholders = set()
-    last_placeholder = None
+    first_non_placeholder = next(iter(graph.nodes))
     for node in graph.nodes:
         if node.op != "placeholder":
+            first_non_placeholder = node
             break
-        last_placeholder = node
-        placeholders.add(node)
-    if last_placeholder is None:  # no placeholders, just insert before first node
-        last_placeholder = next(iter(graph.nodes))
+        else:
+            placeholders.add(node)
+
+    def _is_intermediate_tensor_shape_call(node: fx.Node) -> bool:
+        """
+        If a size/stride/storage offset call on an intermediate tensor,
+        we can try to compute the value from input shapes instead.
+        """
+        val = None
+        if (
+            node.target in (
+                torch.ops.aten.sym_size.int,
+                torch.ops.aten.sym_stride.int,
+                torch.ops.aten.storage_offset.default,
+            )
+            and node.args[0].op != "placeholder"  # type: ignore[union-attr]
+        ):  # export: sym_size.int(node, dim)
+            val = _get_sym_val(node)
+        if (
+            node.target == operator.getitem
+            and node.args[0].target in (
+                "size",
+                "stride",
+                "storage_offset",
+            )
+            and node.args[0].args[0].op != "placeholder"
+        ):  # dynamo export: getitem(size(node), dim)
+            val = _get_sym_val(node)
+        return (
+            val is not None
+            and not isinstance(val, sympy.Number)
+        )
 
     # Identify what symbols we need to reify.  This isn't strictly needed
     # but helps reduce churn on the graph
```

Comment:
It might be simpler to just not identify what symbols we need to reify, and unconditionally reify all of them

Reply:
I see, I think dynamo more nicely handles dead code than export after this pass happens, export will leave these unused nodes in the graph. But I'll add this at the end to delete unused reified symbols, this should be simpler than tracking needed_symbols now:
```
    # delete unused reified symbols
    for expr, proxy in symbol_to_proxy.items():
        if (
            isinstance(expr, sympy.Symbol)
            and not proxy.node.users
        ):
            log.debug("deleting unused reified symbol for %s", expr)
            gm.graph.erase_node(proxy.node)
```